Add new turtle test node based directly on NodeBase

gavin-norman-sociomantic opened the pull request:
(Currently based on #350. Only the last commit is relevant.)

Commit:
Remove unused import

gavin-norman-sociomantic commented:
I believe the separation of turtle node from the actual node implementation is a historical artefact from the days when the fake nodes lived partially in turtle. It seems to be unnecessary now.

gavin-norman-sociomantic commented:
https://github.com/sociomantic-tsunami/dlsproto/pull/101 gives a rough example of how this change simplifies the code for a fake node.

mihails-strasuns commented:
Hm, this doesn't look like any code I have written and turtle definitely doesn't care so I'd say "whatever fits you best" :)

gavin-norman-sociomantic commented:
It was derived from code in turtle that you wrote originally.

mihails-strasuns commented:
I pretty sure that it was you who wrote that code in turtle originally :) I copied it to swarm with almost no changes when splitting things.

mihails-strasuns commented on src/turtle/env/model/TestNode.d:
```diff
@@ -0,0 +1,249 @@
+        Notes:
+            1. Restarting the node *does not* clear any data in its storage
+               engine. To do that, call clear().
+            2. You must call stop() first, before calling restart().
```
This is extremely weird tbh. If you need to call `stop` first, it is just `start`, not `restart`.

gavin-norman-sociomantic replied:
Agreed. I have no memory of why we did it like this.

mihails-strasuns commented on src/turtle/env/model/TestNode.d:
```diff
@@ -0,0 +1,249 @@
+
+    final public void stop ( )
+    {
+        verify(this.state == State.Running, "Node is not running");
```
You need to unregister it from turtle here

gavin-norman-sociomantic commented:
>I pretty sure that it was you who wrote that code in turtle originally :) I copied it to swarm with almost no changes when splitting things.

Could be :) Mainly I'm interested in your opinion on the concept, and whether I'm missing some subtlety of turtle.

mihails-strasuns commented:
AFAIR turtle only cares that nodes stop printing messages once `unregister` is called and everything else is completely up to you.

gavin-norman-sociomantic commented:
Adapted with working restart. (Still super rough.)

Commit:
Move construction of INodeBase's select listeners into abstract methods

References to the listeners are required by INodeBase, but these were
previously passed into the ctor from outside, rather than being owned and
controlled by this class. This, combined with the lack of support in ocean
for reopening a shutdown listener, made it impossible to implement a method
to restart the node's listeners.

(The ideal solution, of course, would be to add the missing restart support
to the select listener in ocean, but this will require a major swarm
release.)

Commit:
Add NodeBase.restartListeners

(As a counterpart to the existing stopListener.)

Commit:
Rename NodeBase.stopListener -> stopListeners

There's more than one listener now.

Commit:
Add turtle.env.model.TestNode

This is intended as a replacement of the existing turtle.env.model.Node.
The reason for the new class is that the turtle/swarm node framework is
currently fairly complicated, with several different classes required to
implement a fake node. The new TestNode derives directly from NodeBase (the
standard base class for swarm nodes), whereas the older Node does not. This
means that to implement a fake node using TestNode, one does not have to
separately write a class deriving from NodeBase.

Commit:
Adapt the turtle env node test to use the new TestNode

gavin-norman-sociomantic commented:
Tidied up.

gavin-norman-sociomantic commented:
D2 fixes.

gavin-norman-sociomantic commented:
Still TODO:
* Release notes (new feature / deprecation).
* Decide whether the changes to `NodeBase` are breaking, and whether there's a non-breaking way of doing them.

nemanja-boric-sociomantic commented on src/swarm/node/model/NeoNode.d:
```diff
@@ -199,10 +199,7 @@ public abstract class INodeBase : INode, INodeInfo
-                  lazy ISelectListener listener,
-                  ISelectListener neo_listener = null,
-                  ISelectListener unix_listener = null )
+    public this ( NodeItem node, ConnectionSetupParams conn_setup_params )
```
docs

nemanja-boric-sociomantic commented on src/turtle/env/model/Node.d:
```diff
@@ -34,6 +34,7 @@ import ocean.core.Verify;
 *******************************************************************************/
 
+deprecated("Use turtle.env.model.TestNode instead")
 public abstract class Node ( NodeType, istring id ) : ITurtleEnv
```
This needs a deprecation note, since it's used by protos (outside swarm).

nemanja-boric-sociomantic commented on src/swarm/node/model/NeoNode.d:
```diff
@@ -327,7 +327,20 @@ public abstract class INodeBase : INode, INodeInfo
 
     ***************************************************************************/
 
+    deprecated("Use `stopListeners()` instead")
```
Deprecation note.

nemanja-boric-sociomantic commented on src/turtle/env/model/TestNode.d:
```diff
@@ -0,0 +1,273 @@
+
+    static this ( )
+    {
+        log = Log.lookup("testnode");
```
Hmmm. Is there a way for the fake node implementation to override the `testnode` log name? It's useful when testing with multiple types of test nodes.

gavin-norman-sociomantic replied:
Yeah, I'll change it to one logger per instance.

nemanja-boric-sociomantic commented on src/turtle/env/model/TestNode.d:
```diff
@@ -0,0 +1,273 @@
+        ConnectionSetupParams conn_setup_params, Options options, int backlog )
+    {
+        verify(!already_created, "Can only have one " ~ idup(this.id) ~
+            " per turtle test app");
```
This is a case for `assert`?

gavin-norman-sociomantic replied:
Why?

nemanja-boric-sociomantic commented on src/turtle/env/model/TestNode.d:
```diff
@@ -0,0 +1,273 @@
+
+        this.register(theScheduler.epoll);
+        if ( Task.getThis() )
+            theScheduler.processEvents();
```
Is there a case where this is not running in task? With the turtle, I don't think so, but I might be mistaken.

nemanja-boric-sociomantic replied:
(note that you require that `theScheduler.epoll` is something - perhaps that's worth verifying as well?)

nemanja-boric-sociomantic commented on src/turtle/env/model/TestNode.d:
```diff
@@ -0,0 +1,273 @@
+
+    ***************************************************************************/
+
+    abstract public void clear ( );
```
Hm, what's the reason for having this abstract method?

gavin-norman-sociomantic replied:
True, it's not required at this level.

nemanja-boric-sociomantic commented on integrationtest/turtleenv/main.d:
```diff
@@ -12,103 +12,246 @@
+            ( Client.Neo.Get.Notification, Const!(Client.Neo.Get.Args) ) { });
+        if ( !ok )
+            return false;
+
```
`if (value[] != "hello") return false;`

nemanja-boric-sociomantic commented:
Needs rebase now.

gavin-norman-sociomantic commented:
Rebased.

nemanja-boric-sociomantic commented:
Hmmmm

```
./src/swarm/node/model/Node.d(702): Error: class swarm.node.model.Node.TestNode interface function INode.restartListeners isn't implemented
./src/swarm/node/model/Node.d(702): Error: class swarm.node.model.Node.TestNode interface function INode.restartListeners isn't implemented
./src/swarm/node/model/Node.d(702): Error: class swarm.node.model.Node.TestNode interface function INode.restartListeners isn't implemented```

nemanja-boric-sociomantic commented:
I also left a review on this PR :-)

gavin-norman-sociomantic commented:
Ohh shit. I changed `INodeBase`, but only adapted the neo node classes that implement that interface :/

gavin-norman-sociomantic commented:
Updated.

mihails-strasuns commented:
@gavin-norman-sociomantic was there anything left for this PR from your side? It looks fine to me but there is a plenty of neo stuff I have no idea about :)

gavin-norman-sociomantic commented:
The PR looks sensible to me, in isolation. I think it was waiting on trying to adapt one of the proto fake nodes, to confirm that it works (and is helpful) in practice.

gavin-norman-sociomantic commented:
Ah, this appears to be a test of adapting a *proto to use this: https://github.com/sociomantic-tsunami/dlsproto/pull/101
